### PR TITLE
added basic statistics oneAPI interface

### PR DESCRIPTION
**Pull request by makart19**

### CODE REVIEW

File: cpp/oneapi/dal/algo/basic_statistics/backend/basic_statistics_interop.hpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    if ((res_op.test(res_min_max) && res_op.get_mask() ^ res_min_max.get_mask()) ||
+        (res_op.test(res_mean_varc) && res_op.get_mask() ^ res_mean_varc.get_mask())) {
+        return daal_lom::estimatesAll;
+    }
+    else if (res_op.test(res_min_max)) {
+        return daal_lom::estimatesMinMax;
+    }
+    else if (res_op.test(res_mean_varc)) {
+        return daal_lom::estimatesMeanVariance;
+    }
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/basic_statistics/backend/basic_statistics_interop.hpp:47]: Can we simplify such conditions?

File: cpp/oneapi/dal/algo/basic_statistics/backend/cpu/compute_kernel_dense.cpp [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+template <typename Float, daal::CpuType Cpu>
+using daal_lom_kernel_t =
+    daal_lom::internal::LowOrderMomentsBatchKernel<Float, daal_lom::defaultDense, Cpu>;
+
+template <typename Method>
+constexpr daal_lom::Method get_daal_method() {
+    daal_lom::Method m = daal_lom::defaultDense;
+
+    return m;
+}
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/basic_statistics/backend/cpu/compute_kernel_dense.cpp:48]:
```suggestion
template <typename Method>
constexpr daal_lom::Method get_daal_method() {
    return daal_lom::defaultDense;
}
```

File: cpp/oneapi/dal/algo/basic_statistics/backend/cpu/compute_kernel_dense.cpp [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        daal_result.allocate<Float>(&daal_input, &daal_parameter, get_daal_method<method_t>()));
+
+    interop::status_to_exception(
+        interop::call_daal_kernel<Float, daal_lom_kernel_t>(ctx,
+                                                            daal_data.get(),
+                                                            &daal_result,
+                                                            &daal_parameter));
+
+    auto result = get_result<Float, task_t>(daal_result);
+    result.set_result_options(desc.get_result_options());
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/basic_statistics/backend/cpu/compute_kernel_dense.cpp:72]:
```suggestion
    auto result = get_result<Float, task_t>(daal_result).set_result_options(desc.get_result_options());
```

File: cpp/oneapi/dal/algo/basic_statistics/backend/cpu/compute_kernel_dense.cpp [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/basic_statistics/backend/cpu/compute_kernel_dense.cpp:73]: Why doesn't it set result tables?

File: cpp/oneapi/dal/algo/basic_statistics/backend/gpu/compute_kernel_dense_dpc.cpp [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+template <typename Float>
+using daal_lom_kernel_t =
+    daal_lom::oneapi::internal::LowOrderMomentsBatchKernelOneAPI<Float, daal_lom::defaultDense>;
+
+template <typename Method>
+constexpr daal_lom::Method get_daal_method() {
+    daal_lom::Method m = daal_lom::defaultDense;
+
+    return m;
+}
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/basic_statistics/backend/gpu/compute_kernel_dense_dpc.cpp:47]: Please do it in one line

File: cpp/oneapi/dal/algo/basic_statistics/backend/cpu/compute_kernel_dense.cpp (hunk 4 above)
Comment (makart19) [reply cpp/oneapi/dal/algo/basic_statistics/backend/cpu/compute_kernel_dense.cpp:73]: they are set in get_result which is defined in interop layer

File: cpp/oneapi/dal/algo/basic_statistics/backend/basic_statistics_interop.hpp (hunk 1 above)
Comment (makart19) [reply cpp/oneapi/dal/algo/basic_statistics/backend/basic_statistics_interop.hpp]: simplified, via adding ~ operator for result_option_id

File: cpp/oneapi/dal/algo/basic_statistics/backend/cpu/compute_kernel_dense.cpp (hunk 2 above)
Comment (makart19) [reply cpp/oneapi/dal/algo/basic_statistics/backend/cpu/compute_kernel_dense.cpp]: fixed

File: cpp/oneapi/dal/algo/basic_statistics/backend/cpu/compute_kernel_dense.cpp (hunk 3 above)
Comment (makart19) [reply cpp/oneapi/dal/algo/basic_statistics/backend/cpu/compute_kernel_dense.cpp]: fixed

File: cpp/oneapi/dal/algo/basic_statistics/backend/gpu/compute_kernel_dense_dpc.cpp (hunk 5 above)
Comment (makart19) [reply cpp/oneapi/dal/algo/basic_statistics/backend/gpu/compute_kernel_dense_dpc.cpp]: fixed

File: cpp/oneapi/dal/algo/basic_statistics/compute_types.cpp [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+
+template <typename Task>
+const result_option_id& compute_result<Task>::get_result_options() const {
+    return impl_->options;
+}
+
+template <typename Task>
+void compute_result<Task>::set_min_impl(const table& value) {
+    impl_->min = value;
+}
```
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/algo/basic_statistics/compute_types.cpp:157]: Setters should also check if valid result options are set (the same behavior as getter). You can check it in k-NN in master

File: cpp/oneapi/dal/algo/basic_statistics/compute_types.cpp [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+
+template <typename Task>
+void compute_result<Task>::set_max_impl(const table& value) {
+    impl_->max = value;
+}
```
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/algo/basic_statistics/compute_types.cpp:162]: same comment for all set methods

File: cpp/oneapi/dal/algo/basic_statistics/compute_types.cpp (hunk 6 above)
Comment (makart19) [reply cpp/oneapi/dal/algo/basic_statistics/compute_types.cpp]: thanks, fixed

File: cpp/oneapi/dal/algo/basic_statistics/compute_types.cpp (hunk 7 above)
Comment (makart19) [reply cpp/oneapi/dal/algo/basic_statistics/compute_types.cpp]: thanks, fixed

**Review by KulikovNikita — APPROVED** [review]

File: cpp/oneapi/dal/algo/basic_statistics/compute_types.hpp [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+class compute_result : public base {
+    static_assert(detail::is_valid_task_v<Task>);
+
+public:
+    using task_t = Task;
+
+    /// Creates a new instance of the class with the default property values.
+    compute_result();
+
+    /// A $n1 \\times n2$ table with the minimum result.
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/basic_statistics/compute_types.hpp:80]: Misleading doxygen for all functions, should be `n \\times 1`. Need to mention that each j-th element is minimum over all observations for the j-th feature. The same for other results.

File: cpp/oneapi/dal/algo/basic_statistics/compute_types.hpp [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+        return *this;
+    }
+
+    auto& set_variation(const table& value) {
+        set_variation_impl(value);
+        return *this;
+    }
+
+    /// Result options that indicates availability of the properties
+    /// @remark default = default_result_options<Task>
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/basic_statistics/compute_types.hpp:171]: Probably, this is error in the source from which the comment was copied. There is no `default_result_options<Task>` in the public API. Need to be fixed as soon as possible to not let it spread out across all code base.

**Review by rlnx — COMMENTED** [review]
Please do not forget about RST documentation, the corresponding task should be planned shortly.

File: cpp/oneapi/dal/algo/basic_statistics/compute_types.hpp (hunk 8 above)
Comment (makart19) [reply cpp/oneapi/dal/algo/basic_statistics/compute_types.hpp]: fixed

File: cpp/oneapi/dal/algo/basic_statistics/compute_types.hpp (hunk 9 above)
Comment (makart19) [reply cpp/oneapi/dal/algo/basic_statistics/compute_types.hpp]: fixed here, and will update in KNN.

**Pull request merged**